aten/src/ATen/core/function_schema_inl.h

```diff
@@ -74,6 +103,82 @@ inline std::string FunctionSchema::formatTypeMismatchMsg(
       *this);
 }
 
+inline bool FunctionSchema::isBackwardCompatibleWith(
+    const FunctionSchema& old,
+    std::ostream* why_not) const {
+  if (!(name() == old.name()
+        && overload_name() == old.overload_name()
+        // we are conservative on is_vararg and is_varret,
+        // since they are only used by internal operators
+        && is_vararg() == old.is_vararg()
+        && is_varret() == old.is_varret()
+        && returns().size() == old.returns().size()
+        && arguments().size() >= old.arguments().size())) {
+    return false;
+  }
+  for (size_t i = 0; i < returns().size(); ++i) {
+    if (!returns().at(i).isBackwardCompatibleWith(
+          old.returns().at(i), true /* is_return */, why_not)) {
+      return false;
+    }
+  }
+  std::vector<const Argument*> args, old_args;
+  std::unordered_map<std::string, const Argument*> kwargs, old_kwargs;
+  auto split_func = [](const std::vector<Argument>& arguments,
+      std::vector<const Argument*>* positionals,
+      std::unordered_map<std::string, const Argument*>* nameds) {
+    for (const Argument& arg : arguments) {
+      if (!arg.kwarg_only()) {
+        positionals->emplace_back(&arg);
+      }
+      nameds->emplace(arg.name(), &arg);
+    }
+  };
+  split_func(arguments(), &args, &kwargs);
+  split_func(old.arguments(), &old_args, &old_kwargs);
+  if (old_args.size() > args.size()) {
+    return false;
+  }
+  for (size_t i = 0; i < old_args.size(); ++i) {
+    if (!args.at(i)->isBackwardCompatibleWith(
+          *old_args.at(i),
+          false /* is_return */,
+          why_not)) {
+      return false;
+    }
+  }
+  for (size_t i = old_args.size(); i < args.size(); ++i) {
+    if (!args.at(i)->default_value()) {
+      auto it = old_kwargs.find(args.at(i)->name());
+      if (it == old_kwargs.end() ||
+          !args.at(i)->isBackwardCompatibleWith(
+            *it->second,
+            false /* is_return */,
+            why_not)) {
+        return false;
+      }
+    }
+  }
+  for (auto& kv : old_kwargs) {
```

Comment:
This is non-deterministic iteration over an unordered map. While it shouldn't change the result it can make debugging harder since it any bugs in this pathway will inherit the non-determinism.

Reply:
Changed to map, the number of args should be small, so I think it's acceptable.